File: src/sr_rngoutputshim.c
#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h>
#include "sr_config.h"

int main(int argc, char** argv) {
    char buffer[_SSRNG_BUFSIZE];
    uint64_t seed = 0;
    uint64_t buf_received = 0;
    char cmd[4096] = "";
    if(fread(&seed, sizeof(uint64_t), 1, stdin) == 0) {
        fprintf(stderr, "srt_rngoutput: error reading stdin");
    }
    sprintf(cmd, "RNG_output %s inf %016llx", argv[1], seed);
    FILE* child = popen(cmd, "r");
    if(child == NULL) {
        fprintf(stderr, "srt_rngoutput: error cannot open %s", cmd);
    }
    while(1) { 
        buf_received = fread(buffer, sizeof(char), _SSRNG_BUFSIZE, child);
        if(buf_received != _SSRNG_BUFSIZE) {
            fprintf(stderr, "srt_rngoutput: error reading child. (%llu,%i)", buf_received, _SSRNG_BUFSIZE);
            break;
        }
        if(buf_received != fwrite(buffer, sizeof(char), _SSRNG_BUFSIZE, stdout)){
            fprintf(stderr, "srt_rngoutput: error writing stdout. (%llu,%i)", buf_received, _SSRNG_BUFSIZE);
            break;
        }
    }
    return EXIT_SUCCESS;
}

File: src/bin/sr_rngoutputshim.rs
use anyhow::{bail, Context, Result};
use std::io::{self, Read, Write};
use std::process::{Command, Stdio};
use softrng::BUFFER_SIZE;

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        eprintln!("sr_rngoutputshim: missing RNG_output argument");
        std::process::exit(1);
    }
    let rng_arg = &args[1];

    // Read 8-byte seed from stdin
    let mut seed_bytes = [0u8; 8];
    io::stdin()
        .read_exact(&mut seed_bytes)
        .context("sr_rngoutputshim: error reading stdin")?;
    let seed = u64::from_le_bytes(seed_bytes);

    // Build command: RNG_output <arg> inf 0123456789abcdef
    let cmd_str = format!("RNG_output {} inf {:016x}", rng_arg, seed);
    let mut child = Command::new("sh")
        .arg("-c")
        .arg(&cmd_str)
        .stdout(Stdio::piped())
        .spawn()
        .with_context(|| format!("sr_rngoutputshim: cannot open {}", cmd_str))?;

    let mut child_stdout = child.stdout.take().expect("failed to capture stdout");
    let mut stdout = io::stdout().lock();
    let mut buffer = vec![0u8; BUFFER_SIZE];

    loop {
        // Read exactly BUFFER_SIZE bytes (as C code expects)
        match child_stdout.read_exact(&mut buffer) {
            Ok(()) => {
                // Write exactly BUFFER_SIZE bytes
                if let Err(e) = stdout.write_all(&buffer) {
                    eprintln!(
                        "sr_rngoutputshim: error writing stdout. ({},{})",
                        BUFFER_SIZE, BUFFER_SIZE
                    );
                    bail!(e);
                }
            }
            Err(e) if e.kind() == io::ErrorKind::UnexpectedEof => {
                // Partial read: C code treats this as error
                eprintln!(
                    "sr_rngoutputshim: error reading child. ({},{})",
                    buffer.len(), BUFFER_SIZE
                );
                break;
            }
            Err(e) => {
                eprintln!(
                    "sr_rngoutputshim: error reading child. ({},{})",
                    0, BUFFER_SIZE
                );
                bail!(e);
            }
        }
    }

    let _ = child.wait();
    Ok(())
}